### PR TITLE
Fix missing sort price and/or product limit/offset when ordering by price on listing page

### PR DESCRIPTION
**Pull request by soulardromain**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Search page: fix sort price using Tools::orderbyPrice function and array_slice for pagination. Category page, manufacturer page, supplier page, new products page, prices drop page: fix sort price using array_slice for pagination. I based on the best sales page.
| Type?         | bug fix
| Category?     | FO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? |  #17212
| How to test?  | Add a specific price on a product and sort any listing page except best sales page and use pagination.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/16698)

### CODE REVIEW

File: classes/Category.php [hunk 1]
```diff
@@ -1049,19 +1051,20 @@ public function getProducts(
...
+        } elseif ($finalOrderBy != 'orderprice') {
+            $sql .= ' ORDER BY ' . (!empty($orderByPrefix) ? $orderByPrefix . '.' : '') . '`' . bqSQL($orderBy) . '` ' . pSQL($orderWay) . '
 			LIMIT ' . (((int) $p - 1) * (int) $n) . ',' . (int) $n;
         }
 
         $result = Db::getInstance(_PS_USE_SQL_SLAVE_)->executeS($sql, true, false);
 
         if (!$result) {
-            return array();
+            return false;
```
Comment (PierreRambaud) [inline classes/Category.php]: Why changing this?

File: classes/Category.php [hunk 2]
```diff
@@ -1049,19 +1051,20 @@ public function getProducts(
 
         if ($random === true) {
             $sql .= ' ORDER BY RAND() LIMIT ' . (int) $randomNumberProducts;
-        } else {
-            $sql .= ' ORDER BY ' . (!empty($orderByPrefix) ? $orderByPrefix . '.' : '') . '`' . bqSQL($orderyBy) . '` ' . pSQL($orderWay) . '
+        } elseif ($finalOrderBy != 'orderprice') {
```
Comment (PierreRambaud) [inline classes/Category.php]: So there is no `ORDER BY`  when ordering by price?

**Comment by PierreRambaud** [comment]
Hi and thanks for your contribution. Can you create an issue to be sure we can reproduce the problem? :)
Kind regards

File: classes/Category.php (hunk 2 above)
Comment (soulardromain) [reply classes/Category.php]: Ordering by price is at line 1065

File: classes/Category.php (hunk 1 above)
Comment (soulardromain) [reply classes/Category.php]: I don't remember, it was probably a mistake when I copied code from the bestsellers page, do I have to commit to cancel this line?

**Comment by soulardromain** [comment]
Hi, i have create an issue : #17212

File: classes/Category.php (hunk 1 above)
Comment (PierreRambaud) [reply classes/Category.php]: Yes, since there is no reason to commit this line :)

Comment (soulardromain) [reply classes/Category.php]: ok done

**Comment by NeOMakinG** [comment]
Hi @soulardromain, there are some conflicts on the PR :)

**Comment by soulardromain** [comment]
Hi @NeOMakinG, conflicts are corrected

File: classes/Product.php [hunk 3]
```diff
@@ -2931,18 +2938,23 @@ public static function getPricesDrop(
...
+        }
 
         $result = Db::getInstance(_PS_USE_SQL_SLAVE_)->executeS($sql);
 
         if (!$result) {
             return false;
         }
 
-        if ($order_by == 'price') {
+        if ($finalOrderBy == 'price') {
```
Comment (NeOMakinG) [inline classes/Product.php]:
```suggestion
        if ($finalOrderBy === 'price') {
```

File: classes/Search.php [hunk 4]
```diff
@@ -426,11 +429,20 @@ public static function find(
 					ON (image_shop.`id_product` = p.`id_product` AND image_shop.cover=1 AND image_shop.id_shop=' . (int) $context->shop->id . ')
 				LEFT JOIN `' . _DB_PREFIX_ . 'image_lang` il ON (image_shop.`id_image` = il.`id_image` AND il.`id_lang` = ' . (int) $id_lang . ')
 				WHERE p.`id_product` ' . $product_pool . '
-				GROUP BY product_shop.id_product
-				' . ($order_by ? 'ORDER BY  ' . $alias . $order_by : '') . ($order_way ? ' ' . $order_way : '') . '
+				GROUP BY product_shop.id_product';
+
+        if ($finalOrderBy != 'price') {
```
Comment (NeOMakinG) [inline classes/Search.php]:
```suggestion
        if ($finalOrderBy !== 'price') {
```

File: classes/Supplier.php [hunk 5]
```diff
@@ -391,18 +393,23 @@ public static function getProducts(
...
+        }
 
         $result = Db::getInstance(_PS_USE_SQL_SLAVE_)->executeS($sql, true, false);
 
         if (!$result) {
             return false;
         }
 
-        if ($orderBy == 'price') {
+        if ($finalOrderBy == 'price') {
```
Comment (NeOMakinG) [inline classes/Supplier.php]:
```suggestion
        if ($finalOrderBy === 'price') {
```

File: classes/Supplier.php [hunk 6]
```diff
@@ -391,18 +393,23 @@ public static function getProducts(
 				WHERE ps.`id_supplier` = ' . (int) $idSupplier . '
 					' . ($active ? ' AND product_shop.`active` = 1' : '') . '
 					' . ($front ? ' AND product_shop.`visibility` IN ("both", "catalog")' : '') . '
-				GROUP BY ps.id_product
-				ORDER BY ' . $alias . pSQL($orderBy) . ' ' . pSQL($orderWay) . '
+				GROUP BY ps.id_product';
+
+        if ($finalOrderBy != 'price') {
```
Comment (NeOMakinG) [inline classes/Supplier.php]:
```suggestion
        if ($finalOrderBy !== 'price') {
```

File: classes/Search.php [hunk 7]
```diff
@@ -426,11 +429,20 @@ public static function find(
...
+            $sql .= ($order_by ? ' ORDER BY  ' . $alias . $order_by : '') . ($order_way ? ' ' . $order_way : '') . '
 				LIMIT ' . (int) (($page_number - 1) * $page_size) . ',' . (int) $page_size;
+        }
+
         $result = $db->executeS($sql, true, false);
 
+        if ($finalOrderBy == 'price') {
```
Comment (NeOMakinG) [inline classes/Search.php]:
```suggestion
        if ($finalOrderBy === 'price') {
```

File: classes/Category.php [hunk 8]
```diff
@@ -991,30 +991,32 @@ public function getProducts(
...
-            $orderyBy = 'name';
-        } elseif ($orderyBy == 'position') {
+            $orderBy = 'name';
+        } elseif ($orderBy == 'position') {
             $orderByPrefix = 'cp';
         }
 
-        if ($orderyBy == 'price') {
-            $orderyBy = 'orderprice';
+        if ($orderBy == 'price') {
```
Comment (NeOMakinG) [inline classes/Category.php]:
```suggestion
        if ($orderBy === 'price') {
```

File: classes/Category.php [hunk 9]
```diff
@@ -991,30 +991,32 @@ public function getProducts(
...
-        } elseif ($orderyBy == 'name') {
+        } elseif ($orderBy == 'name') {
             $orderByPrefix = 'pl';
-        } elseif ($orderyBy == 'manufacturer' || $orderyBy == 'manufacturer_name') {
+        } elseif ($orderBy == 'manufacturer' || $orderBy == 'manufacturer_name') {
             $orderByPrefix = 'm';
-            $orderyBy = 'name';
-        } elseif ($orderyBy == 'position') {
+            $orderBy = 'name';
+        } elseif ($orderBy == 'position') {
```
Comment (NeOMakinG) [inline classes/Category.php]:
```suggestion
        } elseif ($orderBy === 'position') {
```

File: classes/Category.php [hunk 10]
```diff
@@ -991,30 +991,32 @@ public function getProducts(
...
 
         $orderByPrefix = false;
-        if ($orderyBy == 'id_product' || $orderyBy == 'date_add' || $orderyBy == 'date_upd') {
+        if ($orderBy == 'id_product' || $orderBy == 'date_add' || $orderBy == 'date_upd') {
             $orderByPrefix = 'p';
-        } elseif ($orderyBy == 'name') {
+        } elseif ($orderBy == 'name') {
             $orderByPrefix = 'pl';
-        } elseif ($orderyBy == 'manufacturer' || $orderyBy == 'manufacturer_name') {
+        } elseif ($orderBy == 'manufacturer' || $orderBy == 'manufacturer_name') {
```
Comment (NeOMakinG) [inline classes/Category.php]:
```suggestion
        } elseif ($orderBy === 'manufacturer' || $orderBy === 'manufacturer_name') {
```

File: classes/Category.php [hunk 11]
```diff
@@ -991,30 +991,32 @@ public function getProducts(
...
-        $orderyBy = Validate::isOrderBy($orderyBy) ? Tools::strtolower($orderyBy) : 'position';
+        $orderBy = Validate::isOrderBy($orderBy) ? Tools::strtolower($orderBy) : 'position';
         $orderWay = Validate::isOrderWay($orderWay) ? Tools::strtoupper($orderWay) : 'ASC';
 
         $orderByPrefix = false;
-        if ($orderyBy == 'id_product' || $orderyBy == 'date_add' || $orderyBy == 'date_upd') {
+        if ($orderBy == 'id_product' || $orderBy == 'date_add' || $orderBy == 'date_upd') {
             $orderByPrefix = 'p';
-        } elseif ($orderyBy == 'name') {
+        } elseif ($orderBy == 'name') {
```
Comment (NeOMakinG) [inline classes/Category.php]:
```suggestion
        } elseif ($orderBy === 'name') {
```

File: classes/Category.php [hunk 12]
```diff
@@ -991,30 +991,32 @@ public function getProducts(
         }
 
         /** Tools::strtolower is a fix for all modules which are now using lowercase values for 'orderBy' parameter */
-        $orderyBy = Validate::isOrderBy($orderyBy) ? Tools::strtolower($orderyBy) : 'position';
+        $orderBy = Validate::isOrderBy($orderBy) ? Tools::strtolower($orderBy) : 'position';
         $orderWay = Validate::isOrderWay($orderWay) ? Tools::strtoupper($orderWay) : 'ASC';
 
         $orderByPrefix = false;
-        if ($orderyBy == 'id_product' || $orderyBy == 'date_add' || $orderyBy == 'date_upd') {
+        if ($orderBy == 'id_product' || $orderBy == 'date_add' || $orderBy == 'date_upd') {
```
Comment (NeOMakinG) [inline classes/Category.php]:
```suggestion
        if ($orderBy === 'id_product' || $orderBy === 'date_add' || $orderBy === 'date_upd') {
```

File: classes/Category.php [hunk 13]
```diff
@@ -1049,8 +1051,8 @@ public function getProducts(
 
         if ($random === true) {
             $sql .= ' ORDER BY RAND() LIMIT ' . (int) $randomNumberProducts;
-        } else {
-            $sql .= ' ORDER BY ' . (!empty($orderByPrefix) ? $orderByPrefix . '.' : '') . '`' . bqSQL($orderyBy) . '` ' . pSQL($orderWay) . '
+        } elseif ($finalOrderBy != 'orderprice') {
```
Comment (NeOMakinG) [inline classes/Category.php]:
```suggestion
        } elseif ($finalOrderBy !== 'orderprice') {
```

File: classes/Category.php [hunk 14]
```diff
@@ -1060,8 +1062,9 @@ public function getProducts(
             return [];
         }
 
-        if ($orderyBy == 'orderprice') {
+        if ($finalOrderBy == 'orderprice') {
```
Comment (NeOMakinG) [inline classes/Category.php]:
```suggestion
        if ($finalOrderBy === 'orderprice') {
```

File: classes/Manufacturer.php [hunk 15]
```diff
@@ -484,18 +486,23 @@ public static function getProducts(
 				WHERE p.`id_manufacturer` = ' . (int) $idManufacturer . '
 				' . ($active ? ' AND product_shop.`active` = 1' : '') . '
 				' . ($front ? ' AND product_shop.`visibility` IN ("both", "catalog")' : '') . '
-				GROUP BY p.id_product
+				GROUP BY p.id_product';
+
+        if ($finalOrderBy != 'price') {
```
Comment (NeOMakinG) [inline classes/Manufacturer.php]:
```suggestion
        if ($finalOrderBy !== 'price') {
```

File: classes/Manufacturer.php [hunk 16]
```diff
@@ -484,18 +486,23 @@ public static function getProducts(
...
+        }
 
         $result = Db::getInstance(_PS_USE_SQL_SLAVE_)->executeS($sql);
 
         if (!$result) {
             return false;
         }
 
-        if ($orderBy == 'price') {
+        if ($finalOrderBy == 'price') {
```
Comment (NeOMakinG) [inline classes/Manufacturer.php]:
```suggestion
        if ($finalOrderBy === 'price') {
```

File: classes/Product.php [hunk 17]
```diff
@@ -2670,8 +2672,10 @@ public static function getNewProducts($id_lang, $page_number = 0, $nb_products =
             WHERE cp.`id_product` = p.`id_product`)');
         }
 
-        $sql->orderBy((isset($order_by_prefix) ? pSQL($order_by_prefix) . '.' : '') . '`' . pSQL($order_by) . '` ' . pSQL($order_way));
-        $sql->limit($nb_products, (int) (($page_number - 1) * $nb_products));
+        if ($finalOrderBy != 'price') {
```
Comment (NeOMakinG) [inline classes/Product.php]:
```suggestion
        if ($finalOrderBy !== 'price') {
```

File: classes/Product.php [hunk 18]
```diff
@@ -2685,8 +2689,9 @@ public static function getNewProducts($id_lang, $page_number = 0, $nb_products =
             return false;
         }
 
-        if ($order_by == 'price') {
+        if ($finalOrderBy == 'price') {
```
Comment (NeOMakinG) [inline classes/Product.php]:
```suggestion
        if ($finalOrderBy === 'price') {
```

**Review by NeOMakinG — CHANGES_REQUESTED** [review]
Please prefer using strict operators :)

Thanks for your contribution

File: classes/Category.php [hunk 19]
```diff
@@ -991,30 +991,32 @@ public function getProducts(
...
+        if ($orderBy === 'price') {
+            $orderBy = 'orderprice';
         }
 
         $nbDaysNewProduct = Configuration::get('PS_NB_DAYS_NEW_PRODUCT');
         if (!Validate::isUnsignedInt($nbDaysNewProduct)) {
             $nbDaysNewProduct = 20;
         }
 
+        $finalOrderBy = $orderBy;
```
Comment (Progi1984) [inline classes/Category.php]:
```suggestion
```

File: classes/Category.php [hunk 20]
```diff
@@ -1049,8 +1051,8 @@ public function getProducts(
 
         if ($random === true) {
             $sql .= ' ORDER BY RAND() LIMIT ' . (int) $randomNumberProducts;
-        } else {
-            $sql .= ' ORDER BY ' . (!empty($orderByPrefix) ? $orderByPrefix . '.' : '') . '`' . bqSQL($orderyBy) . '` ' . pSQL($orderWay) . '
+        } elseif ($finalOrderBy !== 'orderprice') {
```
Comment (Progi1984) [inline classes/Category.php]:
```suggestion
        } elseif ($orderBy !== 'orderprice') {
```

File: classes/Category.php [hunk 21]
```diff
@@ -1060,8 +1062,9 @@ public function getProducts(
             return [];
         }
 
-        if ($orderyBy == 'orderprice') {
+        if ($finalOrderBy === 'orderprice') {
```
Comment (Progi1984) [inline classes/Category.php]:
```suggestion
        if ($orderBy === 'orderprice') {
```

File: classes/Manufacturer.php [hunk 22]
```diff
@@ -445,6 +445,8 @@ public static function getProducts(
             $alias = 'p.';
         }
 
+        $finalOrderBy = $orderBy;
+
```
Comment (Progi1984) [inline classes/Manufacturer.php]:
```suggestion
```

File: classes/Manufacturer.php [hunk 23]
```diff
@@ -484,18 +486,23 @@ public static function getProducts(
 				WHERE p.`id_manufacturer` = ' . (int) $idManufacturer . '
 				' . ($active ? ' AND product_shop.`active` = 1' : '') . '
 				' . ($front ? ' AND product_shop.`visibility` IN ("both", "catalog")' : '') . '
-				GROUP BY p.id_product
+				GROUP BY p.id_product';
+
+        if ($finalOrderBy !== 'price') {
```
Comment (Progi1984) [inline classes/Manufacturer.php]:
```suggestion
        if ($orderBy !== 'price') {
```

File: classes/Manufacturer.php [hunk 24]
```diff
@@ -484,18 +486,23 @@ public static function getProducts(
...
+        }
 
         $result = Db::getInstance(_PS_USE_SQL_SLAVE_)->executeS($sql);
 
         if (!$result) {
             return false;
         }
 
-        if ($orderBy == 'price') {
+        if ($finalOrderBy === 'price') {
```
Comment (Progi1984) [inline classes/Manufacturer.php]:
```suggestion
        if ($orderBy === 'price') {
```

File: classes/Product.php [hunk 25]
```diff
@@ -2620,6 +2620,8 @@ public static function getNewProducts($id_lang, $page_number = 0, $nb_products =
             $order_by = $order_by[1];
         }
 
+        $finalOrderBy = $order_by;
+
```
Comment (Progi1984) [inline classes/Product.php]:
```suggestion
```

File: classes/Product.php [hunk 26]
```diff
@@ -2670,8 +2672,10 @@ public static function getNewProducts($id_lang, $page_number = 0, $nb_products =
             WHERE cp.`id_product` = p.`id_product`)');
         }
 
-        $sql->orderBy((isset($order_by_prefix) ? pSQL($order_by_prefix) . '.' : '') . '`' . pSQL($order_by) . '` ' . pSQL($order_way));
-        $sql->limit($nb_products, (int) (($page_number - 1) * $nb_products));
+        if ($finalOrderBy !== 'price') {
```
Comment (Progi1984) [inline classes/Product.php]:
```suggestion
        if ($order_by !== 'price') {
```

File: classes/Product.php [hunk 27]
```diff
@@ -2685,8 +2689,9 @@ public static function getNewProducts($id_lang, $page_number = 0, $nb_products =
             return false;
         }
 
-        if ($order_by == 'price') {
+        if ($finalOrderBy === 'price') {
```
Comment (Progi1984) [inline classes/Product.php]:
```suggestion
        if ($order_by === 'price') {
```

File: classes/Product.php [hunk 28]
```diff
@@ -2901,6 +2906,8 @@ public static function getPricesDrop(
             $order_by = pSQL($order_by[0]) . '.`' . pSQL($order_by[1]) . '`';
         }
 
+        $finalOrderBy = $order_by;
+
```
Comment (Progi1984) [inline classes/Product.php]:
```suggestion
```

File: classes/Product.php [hunk 29]
```diff
@@ -2931,18 +2938,23 @@ public static function getPricesDrop(
         AND product_shop.`show_price` = 1
         ' . ($front ? ' AND product_shop.`visibility` IN ("both", "catalog")' : '') . '
         ' . ((!$beginning && !$ending) ? ' AND p.`id_product` IN (' . ((is_array($tab_id_product) && count($tab_id_product)) ? implode(', ', $tab_id_product) : 0) . ')' : '') . '
-        ' . $sql_groups . '
-        ORDER BY ' . (isset($order_by_prefix) ? pSQL($order_by_prefix) . '.' : '') . pSQL($order_by) . ' ' . pSQL($order_way) . '
-        LIMIT ' . (int) (($page_number - 1) * $nb_products) . ', ' . (int) $nb_products;
+        ' . $sql_groups;
+
+        if ($finalOrderBy != 'price') {
```
Comment (Progi1984) [inline classes/Product.php]:
```suggestion
        if ($order_by != 'price') {
```

File: classes/Product.php [hunk 30]
```diff
@@ -2931,18 +2938,23 @@ public static function getPricesDrop(
...
+        }
 
         $result = Db::getInstance(_PS_USE_SQL_SLAVE_)->executeS($sql);
 
         if (!$result) {
             return false;
         }
 
-        if ($order_by == 'price') {
+        if ($finalOrderBy === 'price') {
```
Comment (Progi1984) [inline classes/Product.php]:
```suggestion
        if ($order_by === 'price') {
```

File: classes/Search.php [hunk 31]
```diff
@@ -401,6 +401,9 @@ public static function find(
         } elseif (in_array($order_by, ['date_upd', 'date_add'])) {
             $alias = 'p.';
         }
+
+        $finalOrderBy = $order_by;
+
```
Comment (Progi1984) [inline classes/Search.php]:
```suggestion
```

File: classes/Search.php [hunk 32]
```diff
@@ -426,11 +429,20 @@ public static function find(
 					ON (image_shop.`id_product` = p.`id_product` AND image_shop.cover=1 AND image_shop.id_shop=' . (int) $context->shop->id . ')
 				LEFT JOIN `' . _DB_PREFIX_ . 'image_lang` il ON (image_shop.`id_image` = il.`id_image` AND il.`id_lang` = ' . (int) $id_lang . ')
 				WHERE p.`id_product` ' . $product_pool . '
-				GROUP BY product_shop.id_product
-				' . ($order_by ? 'ORDER BY  ' . $alias . $order_by : '') . ($order_way ? ' ' . $order_way : '') . '
+				GROUP BY product_shop.id_product';
+
+        if ($finalOrderBy !== 'price') {
```
Comment (Progi1984) [inline classes/Search.php]:
```suggestion
        if ($order_by !== 'price') {
```

File: classes/Search.php [hunk 33]
```diff
@@ -426,11 +429,20 @@ public static function find(
...
+            $sql .= ($order_by ? ' ORDER BY  ' . $alias . $order_by : '') . ($order_way ? ' ' . $order_way : '') . '
 				LIMIT ' . (int) (($page_number - 1) * $page_size) . ',' . (int) $page_size;
+        }
+
         $result = $db->executeS($sql, true, false);
 
+        if ($finalOrderBy === 'price') {
```
Comment (Progi1984) [inline classes/Search.php]:
```suggestion
        if ($order_by === 'price') {
```

File: classes/Supplier.php [hunk 34]
```diff
@@ -347,6 +347,8 @@ public static function getProducts(
             $alias = 'm.';
         }
 
+        $finalOrderBy = $orderBy;
+
```
Comment (Progi1984) [inline classes/Supplier.php]:
```suggestion
```

File: classes/Supplier.php [hunk 35]
```diff
@@ -391,18 +393,23 @@ public static function getProducts(
 				WHERE ps.`id_supplier` = ' . (int) $idSupplier . '
 					' . ($active ? ' AND product_shop.`active` = 1' : '') . '
 					' . ($front ? ' AND product_shop.`visibility` IN ("both", "catalog")' : '') . '
-				GROUP BY ps.id_product
-				ORDER BY ' . $alias . pSQL($orderBy) . ' ' . pSQL($orderWay) . '
+				GROUP BY ps.id_product';
+
+        if ($finalOrderBy !== 'price') {
```
Comment (Progi1984) [inline classes/Supplier.php]:
```suggestion
        if ($orderBy !== 'price') {
```

File: classes/Supplier.php [hunk 36]
```diff
@@ -391,18 +393,23 @@ public static function getProducts(
...
+        }
 
         $result = Db::getInstance(_PS_USE_SQL_SLAVE_)->executeS($sql, true, false);
 
         if (!$result) {
             return false;
         }
 
-        if ($orderBy == 'price') {
+        if ($finalOrderBy === 'price') {
```
Comment (Progi1984) [inline classes/Supplier.php]:
```suggestion
        if ($orderBy === 'price') {
```

**Review by PierreRambaud — DISMISSED** [review]
LGTM

**Review by Progi1984 — DISMISSED** [review]

**Comment by Progi1984** [comment]
Hello @soulardromain , thanks for your contribution, would you mind to rebase your pull request? Thanks

**Comment by soulardromain** [comment]
Hello @Progi1984, I think I just mishandled...

**Comment by Progi1984** [comment]
@soulardromain Aouch ! Could you recreate a new PR with your fix ? Thansk :)

**Comment by soulardromain** [comment]
Hello @Progi1984, The new PR is #18179

**Comment by Progi1984** [comment]
Ok. I close this one ;)

**Pull request closed without merging**